src/tribler-core/tribler_core/session.py

```diff
@@ -105,14 +104,17 @@ def __init__(self, config, core_test_mode=False):
         self.bootstrap = None
 
         # modules
-        self.ipv8_community_loader = IPv8CommunityLoader()
-        register_default_launchers(self.ipv8_community_loader)
-
+        if not community_loader:
+            community_loader = create_default_loader(config, chant_testnet=chant_testnet)
```

Comment:
Forwarding an entire loader object seems like overkill for the problem you want to solve. Why not just use a boolean like `use_default_launchers` instead of forwarding a `community_loader` object:

```python
self.ipv8_community_loader = IPv8CommunityLoader()
if use_default_launchers:
    register_default_launchers(self.ipv8_community_loader)
```

Reply:
Because it is more flexible. 

With `community_loader` you can change loader for the session without changing `session` (or `ipv8_module_catalog.py`/`comunity_loader.py`) itself.

By introducing `community_loader` I've decreased the count of session's responsibilities.

Or you want something like:

```python
    def __init__(self, config: TriblerConfig, core_test_mode: bool = False,
                 community_loader: IPv8CommunityLoader = None, use_default_launchers: bool = False,
                 chant_testnet: bool = False, trustchain_testnet: bool = False):
        
        if not use_default_launchers:
            community_loader = create_default_loader(config, chant_testnet=chant_testnet)
```
?

Reply:
I fully agree that it is more flexible. However, extra flexibility is not always a good thing, as it makes the code more complex. 

A loader is simply a container that orchestrates launchers. Making a new loader is like subclassing a `list`. There are uses for subclassing the container itself, but I don't see any uses for this in our code base. Even in Gumby, which is arguably using launchers in the most modular fashion, we don't use a custom loader (only custom launchers). For now, at least, this still seems like overkill.

Looking at it again, perhaps even my boolean suggestion is overkill. We could even simply make a method `Session.register_launchers()` that registers the launchers. This could be overwritten by a `Session` subclass.

Reply:
You are correct, fully agree.

My point here, we shouldn't specify a set of launchers in `Session`. We have left open the possibility of easily changing the list of loaded communities by specifying a custom loader, without changing `Session` itself.

It is an application of "Inversion of control" principle.

Probably the version:
```python
    def __init__(self, config: TriblerConfig, core_test_mode: bool = False,
                 community_loader: IPv8CommunityLoader = None,
                 chant_testnet: bool = False, trustchain_testnet: bool = False):
        self.community_loader = community_loader
```
is more strict and сreates fewer dependencies.

Well, it will require a bigger changeset in this PR as I should fix every creation of `Session` instance. 
But that is fine.

Reply:
> We have left open the possibility of easily changing the list of loaded communities by specifying a custom loader, without changing Session itself.

The possibility was already there. A loader is simply a container for launchers that has functionality to load launcher definitions. You can add and remove launchers from a loader. I believe you're missing the following information, causing our disconnect:

**A loader is nothing but a glorified list for launchers.**

You can freely modify this list without making a new list. The launchers are only loaded as a consequence of `Session.start()`, here:

https://github.com/Tribler/tribler/blob/6c25dac44e6c233afa029251e59de3ad8e2e015a/src/tribler-core/tribler_core/session.py#L136-L137

https://github.com/Tribler/tribler/blob/6c25dac44e6c233afa029251e59de3ad8e2e015a/src/tribler-core/tribler_core/session.py#L356

Why I object to your configuration proposal, is because it boils down to the following:

```python
def register_communities(container):
    container.append("1")
    container.append("2")
    container.append("3")

class Old:
    def __init__(self):
        self.community_loader = []
        register_communities(self.community_loader)

class New:
    def __init__(self, community_loader=["1", "2", "3"]):
        self.community_loader = community_loader
```

I hope that makes it more clear. The old implementation adds elements to a container, your proposal supplies a completely new container with prefilled elements. This is why I suggest a boolean option added to the old implementation.

Reply:
Just for note this is exactly what I want:

```python
class New:
    def __init__(self, community_loader=["1", "2", "3"]):
        self.community_loader = community_loader
```